text/0000-extern-existential-type.md

```diff
@@ -0,0 +1,217 @@
+This is not very nice, but exactly like "lang items" and the annotations that exist today,
+so it is nothing worse.
+
+As mentioned in the introduction, code gen can be reasoned about by comparing with generic and inlining).
+We cannot generate for code for generic items until they are instantiated.
+Likewise, imagine that everything that uses an `pub extern existential type` gets an extra parameter,
+and then when the `impl pub extern existential type` is defined, we go back and eliminate that parameter by substituting the actual definition.
+Only then can we generate code.
+This is why from the root crate of compilation (the binary, static library, or other such "final" component), the dependency closure must contain an `impl existential type` for every `pub extern existential type` that's actually used.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`pub extern existential type` can also be formally defined in reference to `existential type`.
+As explained in the guide-level explanation,
+```rust
+(pub <scope>)? extern existential type <name>: <bounds>;
+```
+creates an existential type alias that behaves just like a `use`d `existential type <name>: <bounds>` defined in another modules so it's opaque, while
+```rust
+(pub <scope>)? impl existential type <path> = <type>;
+```
+reveals the definition of the existential type alias at `path` as if it was a regular type alias.
+
+There is a post-hoc coherence rule that every used `pub extern existential type` contains exactly one `impl exisitential type` definition within the entire build-plan of crates.
+"used" here can be roughly defined by tracing all identifiers through their bindings, but should make intuitive sense.
+
+There is nothing preventing private `extern existential type`, or a `impl extern type` in the same module as its `extern existential type`.
+Both these situations make the feature useless and could be linted, but are well defined from the rules above.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Niko Matsakis has expressed concerned about this being abused because singletons are bad.
+I agree singletons are bad, but the connection between existential types and singletons is not obvious at first sight (imagine if we did the same thing with `static`s directly), so I hope this will be sufficiently difficult to abuse.
+Even if we deem this very toxic, better all the use cases I listed above be white-listed and use same mechanism used for consistency, then use a bunch of separate solutions.
+
+Stabilization of many annotations and APIs I call out in the motivation section is imminent, and yes this would delay that a bit if we opted to do this and then rewrite those APIs to use it.
+
+As per the "prior art" section, something like Haskell's backpack is wholly superior.
+But as stabilization of the status quo is imminent, I wanted to pick something easier to implement and closer to existing rust features mentally/pedagogically.
```

Comment:
Readers will be puzzled why a subpar solution is proposed and if the "imminence" motivation is sufficient (I am somewhat on edge about that personally...).